internal/internal.go

```diff
@@ -695,36 +724,131 @@ func getParsedPathAndQuery(path string) ([]interface{}, map[string]interface{},
+					return nil, false, err
+				}
+				if cl != -1 && cl > int64(len(body)) {
+					return nil, true, nil
+				}
+			}
+
 			err := util.Unmarshal([]byte(body), &data)
 			if err != nil {
 				return nil, false, err
 			}
+		} else if strings.Contains(val, "application/grpc") {
+
+			if protoSet == nil {
+				return nil, false, nil
+			}
+
+			if len(rawBody) == 0 {
+				return nil, false, fmt.Errorf("invalid raw body")
+			}
+			if len(parsedPath) <= 1 {
+				return nil, false, fmt.Errorf("invalid parsed path")
+			}
+
+			known, truncated, err := getGRPCBody(rawBody, parsedPath, &data, protoSet)
+			if err != nil {
+				return nil, false, err
+			}
+			if truncated {
+				return nil, true, nil
+			}
+			if !known {
+				return nil, false, nil
+			}
 		}
 	}
 
 	return data, false, nil
 }
 
+func getGRPCBody(in []byte, parsedPath []interface{}, data interface{}, files *protoregistry.Files) (found, truncated bool, _ error) {
+
+	// the first 5 bytes are part of gRPC framing. We need to remove them to be able to parse
+	// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+
+	if len(in) < 5 {
+		return false, false, fmt.Errorf("less than 5 bytes")
+	}
+
+	// Can be 0 or 1, 1 indicates thet the payload is compressed.
+	// The method could be looked up in the request headers, and the
+	// request decompressed; but for now, let's skip it.
+	if in[0] != 0 {
+		return false, false, nil
+	}
+
+	// Note: we're only reading one message, this is the first message's size
+	size := binary.BigEndian.Uint32(in[1:5])
+	if int(size) > len(in)-5 {
+		return false, true, nil // truncated body
+	}
+	in = in[5 : size+5]
+
+	// Note: we've already checked that len(path)>=2
+	svc, err := findService(parsedPath[0].(string), files)
+	if err != nil {
+		return false, false, nil
+	}
+	msgDesc, err := findMessageInputDesc(parsedPath[1].(string), svc)
+	if err != nil {
+		return false, false, nil
+	}
+
+	msg := dynamicpb.NewMessage(msgDesc)
+	if err := proto.Unmarshal(in, msg); err != nil {
+		return true, false, err
+	}
+
+	jsonBody, err := protojson.Marshal(msg)
+	if err != nil {
+		return true, false, err
+	}
+
+	if err := util.Unmarshal([]byte(jsonBody), &data); err != nil {
+		return true, false, err
+	}
+
+	return true, false, nil
+}
+
+func findService(path string, files *protoregistry.Files) (protoreflect.ServiceDescriptor, error) {
+	desc, err := files.FindDescriptorByName(protoreflect.FullName(path))
+	if err != nil {
+		return nil, err
+	}
+	svcDesc, ok := desc.(protoreflect.ServiceDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("could not find service descriptor for path %q", path)
+	}
+	return svcDesc, nil
+}
+
+func findMessageInputDesc(name string, svc protoreflect.ServiceDescriptor) (protoreflect.MessageDescriptor, error) {
+	if method := svc.Methods().ByName(protoreflect.Name(name)); method != nil {
+		if method.IsStreamingClient() {
+			return nil, fmt.Errorf("method %s is a streaming client method, unsupported", method.Name())
```

Comment:
Nit: Let's reword the error message a bit. For example, `streaming client method foo not supported`